static/js/narrow.js

```diff
@@ -234,11 +249,18 @@ exports.activate = function (raw_operators, opts) {
         ui.show_loading_more_messages_indicator();
     }
 
-    // Put the narrow operators in the URL fragment.
-    // Disabled when the URL fragment was the source
-    // of this narrow.
+    // Put the narrow operators in the URL fragment. Disabled when the URL
+    // fragment was the initiating source of this narrow.
     if (opts.change_hash) {
         hashchange.save_narrow(operators);
+    } else if (following_alias) {
+        // We want this to update the hash on page load if someone followed
+        // a link to a renamed stream. Since hashchange will refuse us right now
+        // (it's already tracking a page load), we process on a future tick.
+        // We set the "replace_state" parameter true to avoid polluting history.
+        setTimeout(function () {
+            hashchange.save_narrow(operators, true);
```

Comment:
I see, if it's just the hash that's updating in the next tick, that's pretty invisible and fine.  I just don't want a flicker from the "stream doesn't exist" notice and back.

BTW http://zulip.readthedocs.io/en/latest/hashchange-system.html may be useful reading if you haven't found it, but I 100% agree that the system is thorny and could benefit from a restructuring.

@showell what are your thoughts?  If you have time to take over this review in general, that'd be great.